core/incremental/expr_compiler.rs

```diff
@@ -0,0 +1,441 @@
+                        }
+                        (Value::Float(a), Value::Float(b)) => {
+                            if *b != 0.0 {
+                                Value::Float(a / b)
+                            } else {
+                                Value::Null
+                            }
+                        }
+                        (Value::Null, _) | (_, Value::Null) => Value::Null,
+                        _ => panic!("Type mismatch in trivial expression: {left_val:?} / {right_val:?}. This is a bug in trivial expression validation."),
+                    },
+                    _ => panic!("Unsupported operator in trivial expression: {op:?}"),
+                }
+            }
+        }
+    }
+}
+
+/// Compiled expression that can be executed on row values
+#[derive(Clone)]
+pub struct CompiledExpression {
+    /// The expression executor (trivial or compiled)
+    pub executor: ExpressionExecutor,
+    /// Number of input values expected (columns from the row)
+    pub input_count: usize,
+}
+
+impl std::fmt::Debug for CompiledExpression {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut s = f.debug_struct("CompiledExpression");
+        s.field("input_count", &self.input_count);
+        match &self.executor {
+            ExpressionExecutor::Trivial(t) => s.field("executor", &format!("Trivial({t:?})")),
+            ExpressionExecutor::Compiled(p) => {
+                s.field("executor", &format!("Compiled({} insns)", p.insns.len()))
+            }
+        };
+        s.finish()
+    }
+}
+
+impl CompiledExpression {
+    /// Get the "type" of a trivial expression for type checking
+    /// Returns None if type can't be determined statically
+    fn get_trivial_type(expr: &TrivialExpression) -> Option<&'static str> {
+        match expr {
+            TrivialExpression::Column(_) => None, // Can't know column type statically
+            TrivialExpression::Immediate(val) => match val {
+                Value::Integer(_) => Some("integer"),
+                Value::Float(_) => Some("float"),
+                Value::Text(_) => Some("text"),
+                Value::Null => Some("null"),
+                _ => None,
+            },
+            TrivialExpression::Binary { left, right, .. } => {
+                // For binary ops, both sides must have the same type
+                let left_type = Self::get_trivial_type(left)?;
+                let right_type = Self::get_trivial_type(right)?;
+                if left_type == right_type {
+                    Some(left_type)
+                } else {
+                    None // Type mismatch
+                }
+            }
+        }
+    }
+
+    /// Check if an expression is trivial (columns, immediates, and simple arithmetic)
+    /// Only considers expressions trivial if they don't require type coercion
+    fn is_trivial_expr(expr: &Expr, input_column_names: &[String]) -> Option<TrivialExpression> {
```

Comment:
`is_trivial_expr` suggests it returns a boolean, perhaps `try_get_trivial_expr()`?